### DISCUSSION TITLE
Bug in offset?

### DISCUSSION THREAD

**Discussion by BarbourSmith** [sgenoud/replicad#172] [Modelling help] [answered]
Hello! 

I am playing around with offset and I have two questions 🙃 

First, when I run this code it's telling me there is a bug in the offset code, but I'm not sure why:

```
const { sketchCircle, offsetBlueprint, drawProjection } = replicad;

const main = () => {
  // Create a sketch of a circle
  const circleSketch1 = sketchCircle(20);
  const circleSketch2= sketchCircle(30);

  // Extrude the sketch
  const extrudedCircle1 = circleSketch1.extrude(10);
  const extrudedCircle2 = circleSketch2.extrude(10);

  const together = extrudedCircle2.cut(extrudedCircle1);

  const projection = drawProjection(together, "top").visible

  const offset = projection.offset(2);

  return offset;
};
```

<img width="720" alt="image" src="https://github.com/user-attachments/assets/09aea90f-0cae-4d63-86a8-ada39a1bea99">

Second, I actually had this working before but it would only offset on the outside of the ring, while I would like to offset to the outside and also to the inside. Is there any way to do that?

Thanks!! 🙏

**Comment by sgenoud** [comment, marked as answer]
The offset of a projection is something that is prone to errors - as offsets for Bézier do not work great (as of now).

Could you do the offset on the sketch step?

```js
const { drawCircle } = replicad;

const main = () => {  
  const circle1 = drawCircle(20);
  const circle2 = drawCircle(30);

  const together = circle2.cut(circle1);

  return [
    circle2.offset(2).cut(circle2.offset(-2)),
    {shape: together.translate(100), color: "red"}
  ]
};
```

[direct link](https://studio.replicad.xyz/workbench#code=UEsDBAoAAAAIAM1xb1kw5I8dpQAAABsBAAAHAAAAY29kZS5qc13OSwqDMBAG4H1OMbhKwIqmO6Xd9BilixDHKqSJTCJdiHdvqpI%252BdsM%252F3zy0sz7ADC2p52UgbRAWOAHhaAat2oYxvYqHGmzMuYDTOXJgAFtDr0NV7H1WcFmK5k%252FIX3F8i0SCu2PokaLZdaGnwPfdmyQME1m4xhKScl3nMXApvnxKD1KIfOWz79WIdbpTBFLWGxWQV2Up8viGcVRDRthmSxy5saV5AVBLAQIUAAoAAAAIAM1xb1kw5I8dpQAAABsBAAAHAAAAAAAAAAAAAAAAAAAAAABjb2RlLmpzUEsFBgAAAAABAAEANQAAAMoAAAAAAA%253D%253D)

 I am working on better offsets for Bézier, but for pantograph - which should become the draw level API for replicad.

**Reply by BarbourSmith** [reply]
That's awesome to hear that it's in the works! 

The goal here is to generate g-code for CNC routing (ie we need to offset from around the edge of the shape by 1/2 of the tool width) so we're stuck dealing with 3D shapes. In this simple case we could do it at the sketch level, but ultimately we'll have to be able to handle arbitrary 3D shapes (that are flat of course so they can be cut out).

There's no way to go from a 3D shape back to a sketch, right?

**Reply by sgenoud** [reply]
Does your g-code support bézier curves (as far as I know it probably does not). We could also make a circular approximation for all the curves in the projection and then offset them, as offsets of circles are easy.

**Reply by BarbourSmith** [reply]
That is an excellent observation. Gcode doesn't support Bézier curves, just circles (well arcs really) and lines. Some programs won't even support circles and instead will just break curves down into line segments which is usually a fine approximation.
